### PR TITLE
NEW PROVIDER: deSEC

### LINKED ISSUE #724: Provider Request: deSEC
Hi!

The non-profit DNS Provider deSEC.de from Germany is now out of beta 🥳

Can you please implement their API https://desec.readthedocs.io?

Thanks!

### PR DESCRIPTION
**Pull request by D3luxee**
This should fix #724

### CODE REVIEW

File: providers/desec/desecProvider.go [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+							F: func() error {
+								err := client.createRR(n, domain)
+								if err != nil {
+									return err
+								}
+								return nil
+							},
+						})
+				}
+			}*/
```
Comment (Knight1) [inline providers/desec/desecProvider.go]: remove?

File: providers/desec/desecProvider.go [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+	providers.RegisterDomainServiceProviderType("DESEC", NewDeSec, features)
+}
+
+// Specific stuff below
+// Specific stuff below
+// Specific stuff below
+// Specific stuff below
+// Specific stuff below
+// Specific stuff below
+// Specific stuff below
```
Comment (Knight1) [inline providers/desec/desecProvider.go]: remove?

**Comment by Knight1** [comment]
README is not updated.

File: providers/desec/desecProvider.go [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+	return c, nil
+}
+
+var features = providers.DocumentationNotes{
+	providers.DocDualHost:            providers.Unimplemented(),
+	providers.DocOfficiallySupported: providers.Cannot(),
+	providers.DocCreateDomains:       providers.Can(),
+	providers.CanUseAlias:            providers.Cannot(),
+	providers.CanUseSRV:              providers.Can(),
+	providers.CanUseSSHFP:            providers.Cannot(),
```
Comment (nils-wisiol) [inline providers/desec/desecProvider.go:42]: deSEC supports SSHFP

File: providers/desec/desecProvider.go [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+	recordsToKeep := make([]*models.RecordConfig, 0, len(dc.Records))
+	for _, rec := range dc.Records {
+		if rec.Type == "ALIAS" {
+			// deSEC does not permit ALIAS records, just ignore it
+			printer.Warnf("deSEC does not support alias records\n")
+			continue
+		}
+		if rec.TTL < 3600 {
+			if rec.Type != "NS" {
+				printer.Warnf("deSEC does not support ttls < 3600. Setting ttl of %s type %s from %d to 3600\n", rec.GetLabelFQDN(), rec.Type, rec.TTL)
```
Comment (nils-wisiol) [inline providers/desec/desecProvider.go:127]: deSEC supports lower TTLs if you ask nicely. Maybe there can be a hint in this message to contact support@desec.io?

File: providers/desec/desecProvider.go [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+	providers.DocOfficiallySupported: providers.Cannot(),
+	providers.DocCreateDomains:       providers.Can(),
+	providers.CanUseAlias:            providers.Cannot(),
+	providers.CanUseSRV:              providers.Can(),
+	providers.CanUseSSHFP:            providers.Cannot(),
+	providers.CanUseCAA:              providers.Can(),
+	providers.CanUseTLSA:             providers.Can(),
+	providers.CanUsePTR:              providers.Unimplemented(),
+	providers.CanGetZones:            providers.Can(),
+	providers.CanAutoDNSSEC:          providers.Cannot(),
```
Comment (nils-wisiol) [inline providers/desec/desecProvider.go:47]: deSEC automatically signs, but not sure what AutoDESEC exactly means here.

Comment (D3luxee) [reply providers/desec/desecProvider.go:47]: As far as i understood this CanAutoDNSSEC this would require deSEC to be the registrar not only the authoritive nameserver.

File: providers/desec/desecProvider.go (hunk 3 above)
Comment (D3luxee) [reply providers/desec/desecProvider.go:42]: Okay, this was something i was not sure about. I will update this.

File: providers/desec/desecProvider.go (hunk 1 above)
Comment (D3luxee) [reply providers/desec/desecProvider.go]: This was already removed

**Comment by Knight1** [comment]
Can we implement login with username and password instead of a token the user has to grab from the API?

File: providers/desec/desecProvider.go (hunk 4 above)
Comment (D3luxee) [reply providers/desec/desecProvider.go:127]: I could give a hint in the message, but for dnscontrol it would not make sense.
Maybe i could implement the validation of the ttl based on the response minimum_ttl, this would allow users to utilize this exception.

Could you allow lower ttls for the zone d3luxee.de so i can test this feature?

**Comment by D3luxee** [comment]
> Can we implement login with username and password instead of a token the user has to grab from the API?
tldr; No:

I've tested this / wanted to implement it in this way, but there are some issues.
Performing the login and logout is no problem in terms of rest api calls, but i did not find a suitable place to call the logout function to invalidate the token after usage.
We would need some Close function in the DNSProvider interface, but this would affect all other providers.

There was no good way to implement this, so i've decided to stick with the api token which is even better in terms of security.

File: providers/desec/desecProvider.go (hunk 4 above)
Comment (nils-wisiol) [reply providers/desec/desecProvider.go:127]: Maybe rephrasing the message to "your deSEC account does not support" can give the user a better hint of what's going on?

I did update the minimum ttl for d3luxee.de.

Comment (D3luxee) [reply providers/desec/desecProvider.go:127]: I did implement this check based on the minimum_ttl yesterday and adjusted the message.
The current "uncommited" message looks like this:
`WARNING: Please contact support@desec.io if you need ttls < 3600. Setting ttl of autodiscover.d3luxee.de type A from 60 to 3600`

Do you have any suggestion for the message?

Comment (nils-wisiol) [reply providers/desec/desecProvider.go:127]: lgtm, thanks!

**Comment by peterthomassen** [comment]
Regarding login: 2FA is on one of our top priorities on our roadmap. If dnscontrol handles the login, it would then probably break for users who enabled 2FA.

File: providers/desec/desecProvider.go (hunk 4 above)
Comment (D3luxee) [reply providers/desec/desecProvider.go]: I've tested and pushed the latest changes, dnscontrol will now validate and correct against the minimum_ttl of the specific domain instead of the hardcoded 3600 seconds.

This handles exceptions made by you or changes on the default minimum_ttl.

**Comment by D3luxee** [comment]
> Regarding login: 2FA is on one of our top priorities on our roadmap. If dnscontrol handles the login, it would then probably break for users who enabled 2FA.

Thats true, we should stick to the api token instead of implementing the login. Is the management of api tokens via web interface already on the roadmap? Maybe you could also prioritize this a bit?

**Comment by peterthomassen** [comment]
Yes, it is very much on the roadmap (even before 2FA). We expect to have token management in the GUI in the next few weeks.

File: providers/desec/desecProvider.go (hunk 5 above)
Comment (tlimoncelli) [reply providers/desec/desecProvider.go:48]: AutoDNSSEC means that the DNS provider does all the DNSSEC for you (including holding your private key).

For the first release it is best to disable things if you are unsure about them, then enable them in future releases as they are tested.

File: providers/desec/protocol.go [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+	printer.Printf("If you want to use DNSSec please add the DS record at your registrar using one of the keys:\n")
+	printer.Printf("%+q", dm.Keys)
+	return nil
+}
+
+//upsertRR will create or override the RRSet with the provided resource record.
+func (c *api) upsertRR(rr []resourceRecord, domain string) error {
+	endpoint := fmt.Sprintf("/domains/%s/rrsets/", domain)
+	byt, _ := json.Marshal(rr)
+	if _, err := c.post(endpoint, "PATCH", byt); err != nil {
```
Comment (tlimoncelli) [inline providers/desec/protocol.go:129]: Should this be `PATCH`?  I thought it should be `PUT` but I could be wrong. (Do they both work if the label doesn't previously exist?)

File: providers/desec/protocol.go [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+		return []byte{}, err
+	}
+
+	bodyString, _ := ioutil.ReadAll(resp.Body)
+	// Got error from API ?
+	if resp.StatusCode > 299 {
+		if resp.StatusCode == 429 && retrycnt < 5 {
+			retrycnt++
+			time.Sleep(500 * time.Millisecond)
+			goto retry
```
Comment (tlimoncelli) [inline providers/desec/protocol.go:164]: Handling retries that that is awesome. I need to copy that for digitalocean, which very much needs it.

File: providers/desec/protocol.go [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+		return bodyString, fmt.Errorf("http status %d %s, the api does not provide more information", resp.StatusCode, resp.Status)
+	}
+	return bodyString, nil
+}
+
+func (c *api) post(endpoint, method string, payload []byte) ([]byte, error) {
+	retrycnt := 0
+retry:
+	client := &http.Client{}
+	req, err := http.NewRequest(method, "https://desec.io/api/v1"+endpoint, bytes.NewReader(payload))
```
Comment (tlimoncelli) [inline providers/desec/protocol.go:180]: Please make `https://desec.io/api/v1` a constant since it is used in multiple places.

File: providers/desec/desecProvider.go (hunk 5 above)
Comment (peterthomassen) [reply providers/desec/desecProvider.go:48]: deSEC does handle all DNSSEC aspects, including holding the private key. That's one of the reasons why the project exists. :-)

File: providers/desec/protocol.go (hunk 7 above)
Comment (peterthomassen) [reply providers/desec/protocol.go:164]: You can also look at the `Retry-After:` response header, which has the number of seconds you'll have to wait.

File: providers/desec/protocol.go (hunk 8 above)
Comment (D3luxee) [reply providers/desec/protocol.go]: Done

File: providers/desec/convert.go [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+			}
+		}
+		rcs = append(rcs, rc)
+	}
+
+	return rcs
+}
+
+func recordsToNative(rcs []*models.RecordConfig, origin string) []resourceRecord {
+	// Take a list of RecordConfig and return an equivalent list of ZoneRecords.
```
[the remark this answers is not in the document]
Comment (D3luxee) [reply providers/desec/convert.go]: Yes, thats some leftovers i fixed the comments with the latest commit.

File: providers/desec/convert.go [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+
+	"github.com/StackExchange/dnscontrol/v3/models"
+	"github.com/StackExchange/dnscontrol/v3/pkg/printer"
+)
+
+// nativeToRecord takes a DNS record from deSEC and returns a native RecordConfig struct.
+func nativeToRecords(n resourceRecord, origin string) (rcs []*models.RecordConfig) {
+
+	// deSEC returns all the values for a given label/rtype pair in each
+	// livedns.DomainRecord.  In other words, if there are multiple A
```
[the remark this answers is not in the document]
Comment (D3luxee) [reply providers/desec/convert.go]: Yes, thats some leftovers i fixed the comments with the latest commit.

File: providers/desec/protocol.go (hunk 7 above)
Comment (D3luxee) [reply providers/desec/protocol.go:166]: Thanks.

File: providers/desec/protocol.go (hunk 6 above)
Comment (D3luxee) [reply providers/desec/protocol.go]: Both work almost in the same way, PATCH allowed me to leave out the ttl for deletions. But its better to use PUT and define the full rrsets.

File: docs/_providers/desec.md [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+This provider does not recognize any special metadata fields unique to deSEC.
+
+## Usage
+Example Javascript:
+
+{% highlight js %}
+var REG_NONE = NewRegistrar('none', 'NONE');    // No registrar.
+var deSEC = NewDnsProvider('desec', 'DESEC');  // deSEC
+
+D("example.tld", REG_NONE, DnsProvider(deSEC),
```
[the remark this answers is not in the document]
Comment (D3luxee) [reply docs/_providers/desec.md]: Done

File: providers/desec/desecProvider.go (hunk 5 above)
Comment (tlimoncelli) [reply providers/desec/desecProvider.go:48]: Oh!  That's why its called deSEC!  I get it now!

Comment (D3luxee) [reply providers/desec/desecProvider.go:48]: I'm still not sure if we talk about the same definition of AutoDNSSEC.
deSEC does sign the zone and provide the DS Records to their users, but each user has to transmit this DS Records to their registrar so that they are added on tld level.
And deSEC does sign those zones per default, you do not need to explicitly request it!

For example i got my neuland.tech domain registered at get.tech directly, i can add this domain to deSEC and they sign the zone. But i will have to add the DS Records using the control panel of get.techt manually.

Is this the definition / meaning of this AutoDNSSEC of dnscontrol?

Comment (tlimoncelli) [reply providers/desec/desecProvider.go:48]: I don't use the feature myself (yet!) but I think that's how most providers implement it.

Comment (D3luxee) [reply providers/desec/desecProvider.go:48]: If someone creates the domain / zone via dns control he gets the DS stuff printed, thats how i implemented it in dnscontrol. But its nothing they have to ask for, its the default behavior.

**Comment by D3luxee** [comment]
I've started using deSEC and dnscontrol for my domains now, in my opinion this is ready now.
Finally i got DNSSEC and DANE for my mailserver again: https://dane.sys4.de/smtp/mail.neuland.tech

I'm using A, AAAA, CNAME, MX, CAA, TLSA, TXT and SRV records, and everything works without any issues.

@tlimoncelli @peterthomassen @nils-wisiol do you have any suggestions?

**Comment by tlimoncelli** [comment]
I'm ready to merge.  @peterthomassen @nils-wisiol ?

**Comment by peterthomassen** [comment]
lgtm!

**Comment by peterthomassen** [comment]
Hmm... I just realized that deSEC is missing from the feature table at https://stackexchange.github.io/dnscontrol/provider-list. I had assumed this was autogenerated from the feature flags set in the code (such as AutoDNSSEC etc.). Can I do anything to extend the list?

**Comment by D3luxee** [comment]
I think this file is autogenerated but requires someone to run go generate, i did not notice that its missing in the matrix.

@tlimoncelli can you generate the file?

**Comment by nils-wisiol** [comment]
Thanks @D3luxee @tlimoncelli !

**Comment by tlimoncelli** [comment]
Hi folks!  There are some documentation updates needed. See https://github.com/StackExchange/dnscontrol/issues/730
